# No-goal Contexts

Previous focus structure effects (see notebooks 01 to 06) as well as further exhaustive modeling show consistently the heterogeneity of combined data; by spliting data into goal contexts and no-goal contexts, both effects as well as models improved despite the pruning of observations. 

No-goal contexts were analysed separately to determine whether the mechanisms identified in goal-oriented counterfactuals also appeared in the absence of an explicit consequent goal. Owing to the smaller number of observations, **these analyses are descriptive and intended to provide qualitative comparison rather than inferential conclusions**.

Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### No-Goal Contexts

1. Escape L2 models
2. Transfer Resolution Models
3. Correct Resolution Models

Read Data

In [2]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

### 1. Escape L2 Models

Create escape_2:

Binary Outcome: responses that are not "L2_other": those who managed to escape L2_other by choosing either a *correct* response or an *L1_transfer* response.

In [3]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

Gather no-goal context data

In [4]:
no_goal = subgroup_theoretical[subgroup_theoretical["Goal_Type"]== "no_goal"].copy()

#### Counts and Proportions for no-goal contexts

Escape_L2 responses count for no_goals:

In [5]:
no_goal_escape_L2_counts = pd.crosstab(
    no_goal["escape_L2"],
    no_goal["Response_Full"],
    margins= True)

no_goal_escape_L2_counts

Response_Full,L1_transfer,L2_other,correct,missing_response,All
escape_L2,,,,,
0,0,12,0,0,12
1,34,0,18,2,54
All,34,12,18,2,66


Escape_L2 responses proportions for no-goal contexts:

In [6]:
no_goal_escape_L2_props = no_goal["escape_L2"].value_counts(normalize = True)
no_goal_escape_L2_props

escape_L2
1    0.818182
0    0.181818
Name: proportion, dtype: float64

From 66 observations in the no_goal group, 81.8% of responses managed to escape L2_other options. This is a similar proportion to the total proportion of observations that managaged to escape the L2 landscape for goal contexts (80.3%)

Counts and Proportions by Condition

In [7]:
no_goal_counts_condition = pd.crosstab(
    [no_goal["Agent"],no_goal["Focus"]],
    no_goal["escape_L2"],
    margins = True)

no_goal_counts_condition

escape_L2     0   1  All
Agent Focus             
0     I       3  21   24
      They    5  13   18
1     I       2  14   16
      They    2   6    8
All          12  54   66

In [8]:
no_goal_props_condition = pd.crosstab(
    [no_goal["Agent"],no_goal["Focus"]],
    no_goal["escape_L2"],
    normalize= "index")

no_goal_props_condition

escape_L2           0         1
Agent Focus                    
0     I      0.125000  0.875000
      They   0.277778  0.722222
1     I      0.125000  0.875000
      They   0.250000  0.750000

As opposed to goal-context behaviour, Agency does not seem to play the same part in the process of escaping the L2_other exploration for non-goal contexts. If anything, Focus would need to explain more than Agency.

#### Logistic Models

Import Libraries

In [9]:
import statsmodels.formula.api as smf
import statsmodels.api as sm
from scipy.stats import chi2


Sanity check:

In [10]:
no_goal["escape_L2"].value_counts()

escape_L2
1    54
0    12
Name: count, dtype: int64

**Main Effect Models**

**E12**

escape_L2 ~ Focus

In [11]:
E12 = smf.logit(
    "escape_L2 ~ Focus",
    data=no_goal
    ).fit()

print(E12.summary())

Optimization terminated successfully.
         Current function value: 0.457812
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                   66
Model:                          Logit   Df Residuals:                       64
Method:                           MLE   Df Model:                            1
Date:                Mon, 21 Sep 2026   Pseudo R-squ.:                 0.03444
Time:                        13:21:04   Log-Likelihood:                -30.216
converged:                       True   LL-Null:                       -31.293
Covariance Type:            nonrobust   LLR p-value:                    0.1421
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         1.9459      0.478      4.070      0.000       1.009       2.883
Focus[T.They]    -0.

**E13**

escape_L2 ~ Agent

In [12]:
E13 = smf.logit(
    "escape_L2 ~ Agent",
    data=no_goal
    ).fit()

print(E13.summary())

Optimization terminated successfully.
         Current function value: 0.473694
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                   66
Model:                          Logit   Df Residuals:                       64
Method:                           MLE   Df Model:                            1
Date:                Mon, 21 Sep 2026   Pseudo R-squ.:               0.0009392
Time:                        13:21:04   Log-Likelihood:                -31.264
converged:                       True   LL-Null:                       -31.293
Covariance Type:            nonrobust   LLR p-value:                    0.8084
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.4469      0.393      3.682      0.000       0.677       2.217
Agent          0.1625      0.

- Even though there are only 66 observations to render anything statistically conclusive, escaping the L2 landscape for no-goal context modeling shows to be coherent with EDA. Focus (p-value = 0.142) captures more unique information than Agency (p-value = 0.809) whereas the oppostite hierarchy characterised the goal conditions.


### 2. Transfer Resolution

Create resolved_transfer binary outcome

In [13]:
escapees = subgroup_theoretical[subgroup_theoretical["escape_L2"] == 1].copy()

In [14]:
escapees["resolved_transfer"] = (escapees["Response_Full"] == "L1_transfer").astype(int)

Gather no-goal context data

In [15]:
no_goal_escapees = escapees[escapees["Goal_Type"] == "no_goal"].copy()

In [16]:
no_goal_escapees.shape

(54, 34)

#### Counts and Proportions for no-goal contexts

In [17]:
no_goal_resolved_transfer_counts = pd.crosstab(
    no_goal_escapees["resolved_transfer"],
    no_goal_escapees["Response_Full"],
    margins=True)

no_goal_resolved_transfer_counts

Response_Full,L1_transfer,correct,missing_response,All
resolved_transfer,,,,
0,0,18,2,20
1,34,0,0,34
All,34,18,2,54


In [18]:
no_goal_resolved_transfer_props = no_goal_escapees["resolved_transfer"].value_counts(normalize=True)
no_goal_resolved_transfer_props

resolved_transfer
1    0.62963
0    0.37037
Name: proportion, dtype: float64

From 54 observations no-goal observations that managed to escape the L2-other options, almost 63% were resolved by means of transfer. This is a slightly higher proportion than transfer resolution's proportion for goal contexts (58.4%)

Counts and Proportions by Condition

In [19]:
no_goal_counts_condition = pd.crosstab(
    [no_goal_escapees["Goal_Type"], no_goal_escapees["Agent"], no_goal_escapees["Focus"]],
    no_goal_escapees["resolved_transfer"],
    margins = True)

no_goal_counts_condition

resolved_transfer       0   1  All
Goal_Type Agent Focus             
no_goal   0     I       8  13   21
                They    3  10   13
          1     I       5   9   14
                They    4   2    6
All                    20  34   54

In [20]:
no_goal_props_condition = pd.crosstab(
    [no_goal_escapees["Agent"],no_goal_escapees["Focus"]],
    no_goal_escapees["resolved_transfer"],
    normalize= "index")

no_goal_props_condition

resolved_transfer         0         1
Agent Focus                          
0     I            0.380952  0.619048
      They         0.230769  0.769231
1     I            0.357143  0.642857
      They         0.666667  0.333333

Coherent with EDA, proportions hardly change once agency is introduced when Focus = I; it drastically changes, however, when Focus = They. Therefore, since we are not running interactions, Agency should be explaining more than Focus. 

#### Logistic Models

Sanity Check:

In [21]:
no_goal_escapees["resolved_transfer"].value_counts()

resolved_transfer
1    34
0    20
Name: count, dtype: int64

**Main Effect Models**

In [22]:
RT12 = smf.logit(
    "resolved_transfer ~ Focus",
    data=no_goal_escapees
    ).fit()

print(RT12.summary())

Optimization terminated successfully.
         Current function value: 0.659148
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   54
Model:                          Logit   Df Residuals:                       52
Method:                           MLE   Df Model:                            1
Date:                Mon, 21 Sep 2026   Pseudo R-squ.:               6.712e-06
Time:                        13:21:05   Log-Likelihood:                -35.594
converged:                       True   LL-Null:                       -35.594
Covariance Type:            nonrobust   LLR p-value:                    0.9826
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         0.5261      0.350      1.504      0.133      -0.160       1.212
Focus[T.They]     0.

In [23]:
RT13 = smf.logit(
    "resolved_transfer ~ Agent",
    data=no_goal_escapees
    ).fit()

print(RT13.summary())

Optimization terminated successfully.
         Current function value: 0.651219
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:      resolved_transfer   No. Observations:                   54
Model:                          Logit   Df Residuals:                       52
Method:                           MLE   Df Model:                            1
Date:                Mon, 21 Sep 2026   Pseudo R-squ.:                 0.01204
Time:                        13:21:05   Log-Likelihood:                -35.166
converged:                       True   LL-Null:                       -35.594
Covariance Type:            nonrobust   LLR p-value:                    0.3546
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.7376      0.367      2.012      0.044       0.019       1.456
Agent         -0.5369      0.

- As expected, once no-goal context are analysed on their own, 54 observations are truly not enough to see anything significant. 

- However, consistent with previous descriptive analyses, Agency in this case seems to help better resolution by transfer (p-value= 0.355) as opposed to Focus (p-value= 0.982) whereas the oppostite hierarchy characterised the goal conditions.

### 3. Correct Resolution

Create resolved_correct binary outcome

In [24]:
escapees["resolved_correct"] = (escapees["Response_Full"] == "correct").astype(int)

Gather no-goal context data

In [25]:
no_goal_escapees = escapees[escapees["Goal_Type"] == "no_goal"].copy()

#### Counts and Proportions for no-goal contexts

In [26]:
no_goal_resolved_correct_counts = pd.crosstab(
    no_goal_escapees["resolved_correct"],
    no_goal_escapees["Response_Full"],
    margins=True)

no_goal_resolved_correct_counts

Response_Full,L1_transfer,correct,missing_response,All
resolved_correct,,,,
0,34,0,2,36
1,0,18,0,18
All,34,18,2,54


In [27]:
no_goal_resolved_correct_props = no_goal_escapees["resolved_correct"].value_counts(normalize=True)
no_goal_resolved_correct_props

resolved_correct
0    0.666667
1    0.333333
Name: proportion, dtype: float64

From 54 observations no-goal observations that managed to escape the L2-other options, 33.3% were resolved correctly. The proportion of no-goal observations that were resolved correctly is lower than the proportion of goal context observations that were resolved correctly (40.5%). 

Counts and Proportions by Condition

In [28]:
no_goal_counts_condition = pd.crosstab(
    [no_goal_escapees["Goal_Type"], no_goal_escapees["Agent"], no_goal_escapees["Focus"]],
    no_goal_escapees["resolved_correct"],
    margins = True)

no_goal_counts_condition

resolved_correct        0   1  All
Goal_Type Agent Focus             
no_goal   0     I      15   6   21
                They   10   3   13
          1     I       9   5   14
                They    2   4    6
All                    36  18   54

In [29]:
no_goal_props_condition = pd.crosstab(
    [no_goal_escapees["Agent"],no_goal_escapees["Focus"]],
    no_goal_escapees["resolved_correct"],
    normalize= "index")

no_goal_props_condition

resolved_correct         0         1
Agent Focus                         
0     I           0.714286  0.285714
      They        0.769231  0.230769
1     I           0.642857  0.357143
      They        0.333333  0.666667

The proportion of correct responses increased slightly when agency is introduced in the antecedent when Focus = I; when Focus = They, it increases almost twice as much. However, from opportunity analysis, one subgroup (Agent = 1, Focus = They) showed unusually efficient conversion toward correct responses. Given the limited number of observations, this pattern should be interpreted cautiously.

#### Logistic Models

Sanity Check:

In [30]:
no_goal_escapees["resolved_correct"].value_counts()

resolved_correct
0    36
1    18
Name: count, dtype: int64

**Main Effect Models**

In [31]:
RC12 = smf.logit(
    "resolved_correct ~ Focus",
    data=no_goal_escapees
    ).fit()

print(RC12.summary())

Optimization terminated successfully.
         Current function value: 0.635021
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                   54
Model:                          Logit   Df Residuals:                       52
Method:                           MLE   Df Model:                            1
Date:                Mon, 21 Sep 2026   Pseudo R-squ.:                0.002346
Time:                        13:21:05   Log-Likelihood:                -34.291
converged:                       True   LL-Null:                       -34.372
Covariance Type:            nonrobust   LLR p-value:                    0.6880
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        -0.7802      0.364     -2.143      0.032      -1.494      -0.067
Focus[T.They]     0.

In [32]:
RC13 = smf.logit(
    "resolved_correct ~ Agent",
    data=no_goal_escapees
    ).fit()

print(RC13.summary())

Optimization terminated successfully.
         Current function value: 0.618743
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                   54
Model:                          Logit   Df Residuals:                       52
Method:                           MLE   Df Model:                            1
Date:                Mon, 21 Sep 2026   Pseudo R-squ.:                 0.02792
Time:                        13:21:05   Log-Likelihood:                -33.412
converged:                       True   LL-Null:                       -34.372
Covariance Type:            nonrobust   LLR p-value:                    0.1659
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.0217      0.389     -2.628      0.009      -1.784      -0.260
Agent          0.8210      0.

- Again, once no-goal context are analysed on their own, 54 observations are truly not enough to see anything significant. 

- In any case, Agency in this case also seems to help resolving correctly (p-value = 0.1659) as opposed to Focus (p-value= 0.6880). This also shows that Agency as a predictor for resolving correctly is better than Agency as a predictor for resolving by means of transfer (p = 0.355). Unfortunately, there aren't enough observations to make any claims.

- Looking into agency-focus interactions to explain differences in variation would have been the right way to proceed. 

In summary:

- The no-goal items on their own do not seem to merely introduce statistical noise; they appear to follow a qualitatively different behavioural strategy to goals. This explains the variability we found when analyses were carried out on the combined dataset.

- Across all exploratory analyses, no-goal contexts consistently displayed patterns that contrasted with those observed in goal-context items. This consistency across descriptive summaries and models suggests that no-goal items may reflect a different behavioural regime rather than merely contributing random variability.